pkg/controller/controller.go

```diff
@@ -1362,6 +1376,38 @@ func (p *csiProvisioner) checkCapacity(ctx context.Context, claim *v1.Persistent
 	return false, nil
 }
 
+func (p *csiProvisioner) VolumesInUseByNode(pv *v1.PersistentVolume) (bool, string, error) {
+	if pv.Spec.CSI == nil {
+		return false, "", fmt.Errorf("failed to find CSI PersistentVolume source")
+	}
+
+	uniqueVolHandle, err := generateUniqueCSIVolumeHandle(pv.Spec.CSI)
+	if err != nil {
+		return false, "", err
+	}
+
+	nodes, err := p.nodeLister.List(labels.Everything())
+	if err != nil {
+		return false, "", fmt.Errorf("unexpected error listing nodes from node informer: %v", err)
+	}
+
+	found, nodeName := isVolumeInUse(uniqueVolHandle, nodes)
+	if !found {
+		// Check nodes list from API server.
+		nodes, err := p.client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
```

Comment:
I am trying to reason whether get list from API server is very necessary. 

The only case it is needed is volume was added into VolumeInUse list, but lister 's node has not yet updated? If lister has a delay, I agree checking api server will make it safer. But it does add some cost especially large scale cluster..

Reply:
What's the frequency of node.volumeInUse get updated? I am trying to understand if it is possible that the node.volumeInUse get updated during the volume deletion.

Reply:
> I am trying to reason whether get list from API server is very necessary.
> 
> The only case it is needed is volume was added into VolumeInUse list, but lister 's node has not yet updated? If lister has a delay, I agree checking api server will make it safer. But it does add some cost especially large scale cluster..

@jingxu97 yes I added the API server check because the lister could potentially be out of date. And since its a delete call we should be conservative in making the decision to delete. However O(nodes) is still relatively moderate in the order of thousands for very large clusters. Thoughts?

Reply:
> What's the frequency of node.volumeInUse get updated? I am trying to understand if it is possible that the node.volumeInUse get updated during the volume deletion.

The node status is updated in a separate go thread. And the default periodicity is 10s (kubeCfg.NodeStatusUpdateFrequency.Duration). See [this](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L516)

Yes there are still very edge case scenario where delete can still sneak in.
Say volume is created and deleted very fast, before node status could be updated or say the kubelet is very slow in updating the node status/connectivity issues between kubelet and the API server. This fix acknowledges such known edge cases are not handled.

Reply:
It will be hard to get a definitive answer on this without actual scale testing.

We know the stale device problem is pretty bad and leads to degraded kubelet performance which is hard to diagnose. 

Since we know the extra apiserver calls only happen when volumes are being deleted, I think it will be not too hard to discover any scale problems that arise.

So I'd lean towards keeping the api server check in.

Similarly, for the case when the kubelet doesn't have #98890 we will have extra api server calls that aren't necessary. But it seems hard to avoid that without a lot more version plumbing? So I think what we have here is ok.

Reply:
@bswartz this is also related to the scaling problem you mentioned. ;)

Reply:
With generic ephemeral volumes coming in soon, it's very possible we'll have pvc create/delete as frequent as pods.

Reply:
I think there is another issue here. The fix in - https://github.com/kubernetes/kubernetes/pull/98890 causes CSI volumes that don't have attach/detach to be reported in VolumeInUse status, but the kubelet can start mount operation without waiting for volume to be reported InUse in node. 

This has important side effects for this PR I think. This means that, if user deletes the pv/pvc before it is reported in node's status (i.e before 10s) then we are back to problematic case and then listing all node objects also will not help. 

I think what we might have to do is - https://github.com/kubernetes/kubernetes/blob/4f9317596ccb9fcb37bf183babf3e8640afcca9b/pkg/volume/util/operationexecutor/operation_generator.go#L1398 code we may to special case CSI volumes that support stage operation, so as mount operation can not start until volume is reported in-use by the kubelet. 

This will perhaps avoid about 95% chance of volume not being reported in node.Status